src/components/views/rooms/UserIdentityWarning.tsx

```diff
@@ -105,54 +115,52 @@ export const UserIdentityWarning: React.FC<UserIdentityWarningProps> = ({ room }
-                        // a new user to show a warning for.  If we are already
-                        // displaying a warning, don't change the display.
-                        //
-                        // We have to do this in a callback to
-                        // `setCurrentPrompt` because this function could have
-                        // been called after an `await`, and the `currentPrompt`
-                        // that this function would have may be outdated.
-                        if (!currentPrompt) {
-                            return selectCurrentPrompt();
-                        } else {
-                            return currentPrompt;
-                        }
-                    });
-                }
+            if (!member) return;
+
+            membersNeedingApprovalRef.current.set(userId, member);
+            // We only select the prompt if we are done initialising,
+            // because we will select the prompt after we're done
+            // initialising, and we want to start by displaying a warning
+            // for the user with the smallest ID.
+            if (initialisedRef.current === InitialisationStatus.Completed) {
+                updateCurrentPrompt();
             }
         },
-        [cli, room, selectCurrentPrompt],
+        [cli, room, updateCurrentPrompt],
     );
 
     // Check if the user's identity needs approval, and if so, add them to the
     // membersNeedingApproval map and update the prompt if needed. They will
     // only be added if they are a member of the room.
-    const addMemberIfNeedsApproval = useCallback(
-        async (userId: string, member?: RoomMember): Promise<void> => {
+    const addMembersWhoNeedApproval = useCallback(
+        async (members: RoomMember[]): Promise<void> => {
             const gotVerificationStatusUpdate = gotVerificationStatusUpdateRef.current;
-            const membersNeedingApproval = membersNeedingApprovalRef.current;
 
-            if (gotVerificationStatusUpdate.has(userId)) {
-                // We're already checking their verification status, so we don't
-                // need to do anything here.
-                return;
-            }
-            gotVerificationStatusUpdate.set(userId, false);
-            if (await userNeedsApproval(crypto!, userId)) {
-                if (!membersNeedingApproval.has(userId) && gotVerificationStatusUpdate.get(userId) === false) {
-                    addMemberNeedingApproval(userId, member);
+            const promises: Promise<void>[] = [];
+
+            for (const member of members) {
+                const userId = member.userId;
+                if (gotVerificationStatusUpdate.has(userId)) {
+                    // We're already checking their verification status, so we don't
+                    // need to do anything here.
+                    continue;
                 }
+                gotVerificationStatusUpdate.set(userId, false);
+                promises.push(
```

Comment:
out of interest, was the switch to doing this in parallel meant to fix something, or just an optimisation?

Reply:
Just an optimisation.  I changed the function to take a list of users so that it would update `gotVerificationStatusUpdate` for all the users at once, so that if we get an update for, say, Alice, but we haven't started fetching her update status, we won't try to overwrite the update with a fetched value.  Then I figured it was silly that we were waiting for one to complete before querying the next.